File: 48/가장 긴 증가하는 부분 수열 2/정민우.py
N, A, lis = int(input()), list(map(int, input().split())), []

for a in A:
    if not lis or lis[-1] < a:
        lis.append(a)
    elif a < lis[0]:
        lis[0] = a
    else:
        l, c, r = 0, len(lis) // 2, len(lis)
        while True:
            if lis[c] == a:
                break
            elif lis[c] < a and l == c:
                lis[c + 1] = a
                break
            elif a < lis[c]:
                r, c = c, (c + l) // 2
            elif lis[c] < a:
                l, c = c, (c + r) // 2
print(len(lis))

File: 49/북극곰은 괄호를 찢어/정민우.js
const fs = require("fs");
const [n, input] = fs.readFileSync("/dev/stdin").toString().trim().split("\n");
let [stack, ans] = [[], 0];

for (const i in input) {
  if (!stack.length) {
    stack.push(input[i]);
  } else {
    now = stack.pop();
    if (now === input[i]) {
      stack.push(now);
      stack.push(input[i]);
    }
  }
  if (ans < stack.length) {
    ans = stack.length;
  }
}
stack.length ? console.log(-1) : console.log(ans);
